### DISCUSSION TITLE
Session global fixture parameterization?

### DISCUSSION THREAD

**Discussion by harsszegi** [pytest-dev/pytest#11809] [Q&A]
Hi,

I'm quite new to Pytest and I need to solve the following:
I have an all-test-global conftest.py which has a session fixture, which basically initializes the test environment.
I also have multiple PyTest Python files in different folders testing different submodules of the application. In the files I have loads of Pytests.
I need to be able to parameterize the test environment on a "folder" basis, e.g. every application submodule can have a different test environment parameterization that needs to be passed to the session fixture, so that the fixture code can alter its behavior based upon the received parameter.
I'd like to avoid creating conftest.pys in every submodule folder.
How can it be actually done?
Thanks,

**Comment by nicoddemus** [comment]
Depends on what you mean by "test environment", is that something that is possible to change on a per-test basis in the same session, for example switching some environment variables?

If so, you can define a fixture in your root `conftest.py` file which receives the `request: pytest.FixtureRequest` fixture. Using the fixture, you can get the location of the test requesting this fixture via `request.node.path`, and from there find the "folder" containing the configuration for that test, then apply the configuration for that test.

Something like this:

```python
@pytest.fixture(autouse=True)
def setup_test_environment(request: pytest.FixtureRequest):
    config = get_config_for_path(request.node.path)
    if config:
        apply_env_config(config)
    yield
    # In case you need to undo the configuration somehow:
    if config:
        undo_env_config(config)
```

**Reply by harsszegi** [reply]
It is actually a hardware-device initialization sequence done by the session fixture, I need to be able to alter the initialization sequence on per-session basis, e.g. it can not be altered by test cases. Isn't there a similar solution that doesn't "reverse-engineer" the context information based upon the path containing the given tests, can't I provide that "context information" explicitly somehow? If not, then request.node.path it is, that can also work.
